src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -116,11 +118,18 @@ typesOf bindingGroupType moduleName vals = withFreshSubstitution $ do
     -- Show warnings here, since types in wildcards might have been solved during
     -- instance resolution (by functional dependencies).
     finalSubst <- gets checkSubstitution
-    escalateWarningWhen isHoleError . tell . onErrorMessages (replaceTypes finalSubst) $ w
+    escalateWarningWhen isHoleError . tell . onErrorMessages (replaceTypes False finalSubst) $ w
 
     return inferred
   where
-    replaceTypes subst = onTypesInErrorMessage (substituteType subst)
+    replaceTypes shouldGeneralize subst =
+      (if not shouldGeneralize then runTypeSearch else id)
```

Comment:
Could you please pull this out into a value in the `where` clause and use a guard instead of `if`, for readability?